docs/source/building/variants.rst

```diff
@@ -0,0 +1,911 @@
+
+meta.yaml:
+
+.. code-block:: yaml
+
+   requirements:
+       build:
+           - numpy {{ numpy }}
+       run:
+           - numpy {{ pin_compatible('numpy', pins=['x.x', 'x.x'] }}
+
+
+This would yield a pinning of ``>=1.11,<1.12``
+
+
+You can also pass the maximum version directly. This argument supercedes the
+``pins`` argument and is thus mutually exclusive.
+
+
+.. code-block:: python
+
+    variants = [{'numpy': '1.11'}]
+
+meta.yaml:
+
+.. code-block:: yaml
+
+   requirements:
+       build:
+           - numpy {{ numpy }}
+       run:
+           - numpy {{ pin_compatible('numpy', upper_bound='3.0' }}
+
+
+This would yield a pinning of ``>=1.11,<3.0``
+
+Appending to recipes
+--------------------
+
+
+As of conda-build 3.0, you can add a file named ``recipe_append.yaml`` in the
+same folder as your ``meta.yaml`` file. This file is considered to follow the
+same rules as meta.yaml, except that selectors and Jinja2 templates are not
+(currently) evaluated. That will likely be added in future development.
+
+Any contents in ``recipe_append.yaml`` will add to the contents of meta.yaml.
+List values will be extended, and string values will be concatenated.
+
+
+Partially clobbering recipes
+----------------------------
+
+
+As of conda-build 3.0, you can add a file named ``recipe_clobber.yaml`` in the
+same folder as your ``meta.yaml`` file. This file is considered to follow the
+same rules as meta.yaml, except that selectors and Jinja2 templates are not
+(currently) evaluated. That will likely be added in future development.
+
+Any contents in ``recipe_clobber.yaml`` will replace the contents of meta.yaml.
+This can be useful, for example, for replacing the source URL without copying
+the rest of the recipe into a fork.
+
+
+Differentiating packages built with different variants
+------------------------------------------------------
+
+
+With only a few things supported, we could just add things to the filename, such
+as py27 for python, or np111 for numpy. In the general case, which variants are
+meant to support, this is no longer an option. Instead, part of the recipe is
+hashed using the sha1 algorithm, and that hash is a unique identifier. The
+information that went into the hash is stored with the package, in a file at
+``info/hash_input.json``. Currently, only the first 7 characters of the hash are
+stored. Output package names will keep the pyXY and npXYY for now, but have
+added the 7-character hash. Your package names will look like:
+
+``my-package-1.0-py27h3142afe_0.tar.bz2``
+
+Since conflicts only need to be prevented within one version of a package, we
+think this will be adequate. If you run into hash collisions with this limited
+subspace, please file an issue on the conda-build issue tracker.
+
+The information that goes into this hash is currently defined in conda-build's
+metadata.py module; the _get_hash_contents member function. This function
+captures the following information:
+
```

Comment:
It's a few things.  Relevant code is at https://github.com/conda/conda-build/pull/1585/files#diff-c86934c27c2f25579540ddb447fa45bdR767

1) source, requirements, and build sections of the recipe.  This captures inputs, as well as exact versions of packages present at build time.  The build section has a few options that affect the output, so it's there too.  A few fields in the build section are omitted, so that the hashes are more reproducible.
2) Any other files in the recipe folder (alongside meta.yaml).  Everything but meta.yaml is part of the hash.